src/watchdog.rs

```diff
@@ -0,0 +1,110 @@
+//! Watchdog peripherals
+
+use crate::{
+    hal::watchdog::{Watchdog, WatchdogEnable},
+    stm32::{DBGMCU, IWDG},
+    time::MilliSeconds,
+};
+
+/// Wraps the Independent Watchdog (IWDG) peripheral
+pub struct IndependentWatchdog {
+    iwdg: IWDG,
+}
+
+const LSI_KHZ: u32 = 32;
+const MAX_PR: u8 = 0b110;
+const MAX_RL: u16 = 0xFFF;
+const KR_ACCESS: u16 = 0x5555;
+const KR_RELOAD: u16 = 0xAAAA;
+const KR_START: u16 = 0xCCCC;
+
+impl IndependentWatchdog {
+    /// Creates a new `IndependentWatchDog` without starting it. Call `start` to start the watchdog.
+    /// See `WatchdogEnable` and `Watchdog` for more info.
+    pub fn new(iwdg: IWDG) -> Self {
+        IndependentWatchdog { iwdg }
+    }
+
+    /// Debug independent watchdog stopped when core is halted
+    pub fn stop_on_debug(&self, dbgmcu: &DBGMCU, stop: bool) {
+        #[cfg(any(feature = "stm32l4x1", feature = "stm32l4x2", feature = "stm32l4x3",))]
+        dbgmcu.apb1fzr1.modify(|_, w| w.dbg_iwdg_stop().bit(stop));
+        #[cfg(any(feature = "stm32l4x5", feature = "stm32l4x6"))]
+        dbgmcu.apb1_fzr1.modify(|_, w| w.dbg_iwdg_stop().bit(stop));
+    }
+
+    /// Sets the watchdog timer timout period. Max: 32768 ms
+    fn setup(&self, timeout_ms: u32) {
+        let mut pr = 0;
+        while pr < MAX_PR && Self::timeout_period(pr, MAX_RL) < timeout_ms {
```

Comment:
I think this should be rewritten to allow for the optimizer to see through better.
This does not optimize away for me right now.

Could this be rewritten to not use a loop?

Reply:
The only way I could come up with would be to precalculate `(MAX_RL + 1) * divider / LSI_KHZ` for each divider making a match statement or make a long if .. else if clause.. That would not be reader friendly and quite large. But should optimize i think.

Reply:
Something like this is possible...
```rust
if timeout_ms <= (MAX_PR + 1) * 4 / LSI_KHZ {
    pr = 0b000;
} else if timeout_ms <= (MAX_PR + 1) * 8 / LSI_KHZ {
    pr = 0b001;
} else if timeout_ms <= (MAX_PR + 1) * 16 / LSI_KHZ {
    pr = 0b010;
} else if timeout_ms <= (MAX_PR + 1) * 32 / LSI_KHZ {
    pr = 0b011;
} else if timeout_ms <= (MAX_PR + 1) * 64 / LSI_KHZ {
    pr = 0b100;
} else if timeout_ms <= (MAX_PR + 1) * 128 / LSI_KHZ {
    pr = 0b101;
} else if timeout_ms <= (MAX_PR + 1) * 256 / LSI_KHZ {
    pr = 0b110;
} else {
    pr = 0b110;
}
```

Reply:
Alternative:

```rust
assert!(timeout_ms < (1 << 15));
let pr = match timeout_ms {
    t if t <= (MAX_PR + 1) * 4 / LSI_KHZ => 0b000,
    t if t <= (MAX_PR + 1) * 8 / LSI_KHZ => 0b001,
    t if t <= (MAX_PR + 1) * 16 / LSI_KHZ => 0b010,
    t if t <= (MAX_PR + 1) * 32 / LSI_KHZ => 0b011,
    t if t <= (MAX_PR + 1) * 64 / LSI_KHZ => 0b100,
    t if t <= (MAX_PR + 1) * 128 / LSI_KHZ => 0b101,
    _ => 0x110,
};
```

Reply:
I have gone with Mathias alternative.